### PR TITLE
Integrate DynamicDataReader

### PR DESCRIPTION
**Pull request by sonndinh**

### CODE REVIEW

File: topic_table_model.h [hunk 1]
```diff
@@ -175,6 +179,14 @@ class TopicTableModel : public QAbstractTableModel
      */
     void parseData(const std::shared_ptr<OpenDynamicData> data);
 
+    /// Convert from DDS::TypeKind to CORBA::TCKind. Needed to parse DynamicData.
+    CORBA::TCKind typekind_to_tckind(DDS::TypeKind tk);
+
+    bool check_rc(DDS::ReturnCode_t rc, const char* what);
+
+    /// Parse a DynamicData object into m_data
+    void parseData(const DDS::DynamicData_var data);
```
Comment (mitza-oci) [inline topic_table_model.h:188]:
```suggestion
    void parseData(DDS::DynamicData_var data);
```
OR `const &`

File: topic_monitor.cpp [hunk 2]
```diff
@@ -224,6 +251,30 @@ void TopicMonitor::on_sample_data_received(OpenDDS::DCPS::Recorder*,
...
+    DDS::SampleInfoSeq infos;
+    ddr->take(messages, infos, DDS::LENGTH_UNLIMITED,
+              DDS::ANY_SAMPLE_STATE, DDS::ANY_VIEW_STATE, DDS::ANY_INSTANCE_STATE);
+
+    for (unsigned int i = 0; i < messages.length(); ++i) {
+        if (infos[i].valid_data) {
+            // TODO: Apply content filtering when it's supported.
+            QDateTime dataTime = QDateTime::fromMSecsSinceEpoch(
+            (static_cast<unsigned long long>(infos[i].source_timestamp.sec) * 1000) +
+            (static_cast<unsigned long long>(infos[i].source_timestamp.nanosec) * 1e-6));
```
Comment (mitza-oci) [inline topic_monitor.cpp:271]:
```suggestion
              (static_cast<unsigned long long>(infos[i].source_timestamp.sec) * 1000) +
              (static_cast<unsigned long long>(infos[i].source_timestamp.nanosec) * 1e-6));
```

File: topic_monitor.cpp [hunk 3]
```diff
@@ -224,6 +251,30 @@ void TopicMonitor::on_sample_data_received(OpenDDS::DCPS::Recorder*,
...
+void TopicMonitor::on_data_available(DDS::DataReader_ptr dr)
+{
+    if (m_paused) {
+        return;
+    }
+
+    DDS::DynamicDataReader_var ddr = DDS::DynamicDataReader::_narrow(dr);
+    DDS::DynamicDataSeq messages;
+    DDS::SampleInfoSeq infos;
+    ddr->take(messages, infos, DDS::LENGTH_UNLIMITED,
```
Comment (mitza-oci) [inline topic_monitor.cpp:263]: check return code

File: publication_monitor.h [hunk 4]
```diff
@@ -44,6 +44,10 @@ class PublicationMonitor : public QObject, public virtual GenericReaderListener
 
 private:
 
+    /// Get DynamicType for the corresponding topic
+    bool get_dynamic_type(DDS::DynamicType_var& type, const DDS::BuiltinTopicKey_t & key,
```
Comment (mitza-oci) [inline publication_monitor.h:48]:
```suggestion
    bool get_dynamic_type(DDS::DynamicType_var& type, const DDS::BuiltinTopicKey_t& key,
```

File: dds_data.h [hunk 5]
```diff
@@ -262,6 +271,8 @@ class CommonData
     static std::shared_ptr<OpenDynamicData> copySample(const QString& topicName,
                                        const unsigned int& index);
 
+    static DDS::DynamicData_var copyDynamicSample(const QString& topic_name,
+                                                  const unsigned int index);
```
Comment (mitza-oci) [inline dds_data.h:275]:
```suggestion
                                                  unsigned int index);
```

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: dds_data.cpp [hunk 6]
```diff
@@ -265,6 +267,26 @@ void CommonData::storeSample(const QString& topicName,
 
 } // End CommonData::storeSample
 
+//------------------------------------------------------------------------------
+void CommonData::storeDynamicSample(const QString& topic_name,
+                                    const QString& sample_name,
+                                    const DDS::DynamicData_var sample)
+{
+    m_dynsampleMutex.lock();
```
Comment (simpsont-oci) [inline dds_data.cpp:275]: Generally, use QMutexLocker for scoped locking of QMutex (here and elsewhere).

File: publication_monitor.cpp [hunk 7]
```diff
@@ -142,6 +151,21 @@ void PublicationMonitor::on_data_available(DDS::DataReader_ptr reader)
 
 } // End PublicationMonitor::on_data_available
 
+bool PublicationMonitor::get_dynamic_type(DDS::DynamicType_var& type, const DDS::BuiltinTopicKey_t& key,
+                                          const char* topic_name, const char* type_name)
+{
+    OpenDDS::DCPS::DomainParticipantImpl* participant =
```
Comment (simpsont-oci) [inline publication_monitor.cpp:157]: Not a huge deal, probably, but I would capture the DomainParticipant_var here so the reference count doesn't get decremented until the function exits. Then you can cast var.in() to a DomainParticipantImpl* as a separate variable if you need that for `get_dynamic_type()`.

File: topic_monitor.cpp [hunk 8]
```diff
@@ -29,48 +31,73 @@ TopicMonitor::TopicMonitor(const QString& topicName) :
...
     //store extensibility
     m_extensibility = topicInfo->extensibility;
-
     OpenDDS::DCPS::Service_Participant* service = TheServiceParticipant;
-    DDS::DomainParticipant* domain = CommonData::m_ddsManager ? CommonData::m_ddsManager->getDomainParticipant() : nullptr;
-
-    if (!domain) {
-        std::cerr << "No domain participant" << std::endl;
-        return;
+    DDS::DomainParticipant* participant =
```
Comment (simpsont-oci) [inline topic_monitor.cpp:37]: Same comment here about capturing the DomainParticipant_var to protect the reference count (as above).

File: topic_monitor.cpp [hunk 9]
```diff
@@ -29,48 +31,73 @@ TopicMonitor::TopicMonitor(const QString& topicName) :
...
-        return;
-    }
+        //std::cout << "DEBUG Created typeless topic, now creating recorder" << endl;
+        m_recorder = service->create_recorder(participant,
+                                              m_topic,
+                                              topicInfo->subQos,
+                                              topicInfo->readerQos,
+                                              m_listener);
+        if (!m_recorder) {
+            std::cerr << "Failed to created recorder" << std::endl;
```
Comment (simpsont-oci) [inline topic_monitor.cpp:66]:
```suggestion
            std::cerr << "Failed to created recorder for topic " << topicInfo->name << std::endl;
```

File: topic_monitor.cpp [hunk 10]
```diff
@@ -29,48 +31,73 @@ TopicMonitor::TopicMonitor(const QString& topicName) :
...
-        m_topic,
-        topicInfo->subQos,
-        topicInfo->readerQos,
-        m_listener
-    );
+        DDS::Subscriber_var subscriber = participant->create_subscriber(topicInfo->subQos,
+                                                                        0,
+                                                                        0);
+        if (!subscriber) {
+            std::cerr << "Failed to create subscriber" << std::endl;
```
Comment (simpsont-oci) [inline topic_monitor.cpp:88]:
```suggestion
            std::cerr << "Failed to create subscriber for topic " << topicInfo->name << std::endl;
```

File: topic_monitor.h [hunk 11]
```diff
@@ -74,6 +74,38 @@ class TopicMonitor
         TopicMonitor& m_monitor;
     };
 
+    void on_data_available(DDS::DataReader_ptr);
+
+    class DataReaderListenerImpl : public OpenDDS::DCPS::LocalObject<DDS::DataReaderListener>
```
Comment (simpsont-oci) [inline topic_monitor.h:79]: Resolve windows warnings... maybe needs virtual inheritance here?

**Review by simpsont-oci — CHANGES_REQUESTED** [review]

File: dds_data.h [hunk 12]
```diff
@@ -252,6 +256,11 @@ class CommonData
                             const QString& sampleName,
                             const std::shared_ptr<OpenDynamicData> sample);
 
+    /// Store a new sample represented by a DynamicData object.
+    static void storeDynamicSample(const QString& topic_name,
+                                   const QString& sample_name,
+                                   const DDS::DynamicData_var sample);
```
Comment (mitza-oci) [inline dds_data.h:262]:
```suggestion
                                   DDS::DynamicData_var sample);
```

**Review by mitza-oci — APPROVED** [review]

**Pull request merged**